hotshot-state-prover/src/service.rs

```diff
@@ -0,0 +1,575 @@
+//! A light client prover service
+
+use crate::snark::{generate_state_update_proof, Proof, ProvingKey};
+use anyhow::anyhow;
+use async_std::{
+    io,
+    sync::Arc,
+    task::{sleep, spawn},
+};
+use contract_bindings::light_client::LightClient;
+use displaydoc::Display;
+use ethers::{
+    core::k256::ecdsa::SigningKey,
+    middleware::SignerMiddleware,
+    providers::Http,
+    providers::{Middleware, Provider, ProviderError},
+    signers::{LocalWallet, Signer, Wallet},
+    types::{Address, U256},
+};
+use futures::FutureExt;
+use hotshot_contract_adapter::jellyfish::ParsedPlonkProof;
+use hotshot_contract_adapter::light_client::ParsedLightClientState;
+use hotshot_stake_table::vec_based::StakeTable;
+use hotshot_types::signature_key::BLSPubKey;
+use hotshot_types::traits::stake_table::{SnapshotVersion, StakeTableError, StakeTableScheme as _};
+use hotshot_types::{
+    light_client::{
+        CircuitField, LightClientState, PublicInput, StateKeyPair, StateSignaturesBundle,
+        StateVerKey,
+    },
+    traits::signature_key::SignatureKey,
+};
+use jf_plonk::errors::PlonkError;
+use jf_primitives::pcs::prelude::UnivariateUniversalParams;
+use jf_relation::Circuit as _;
+use std::time::Duration;
+use surf_disco::Client;
+use tide_disco::{error::ServerError, Api};
+use time::Instant;
+use url::Url;
+
+const STAKE_TABLE_CAPACITY: usize = 200;
+
+/// A wallet with local signer and connected to network via http
+pub type L1Wallet = SignerMiddleware<Provider<Http>, LocalWallet>;
+
+/// Configuration/Parameters used for hotshot state prover
+#[derive(Debug, Clone)]
+pub struct StateProverConfig {
+    /// Url of the state relay server (a CDN that sequencers push their Schnorr signatures to)
+    pub relay_server: Url,
+    /// Interval between light client state update
+    pub update_interval: Duration,
+    /// URL of layer 1 Ethereum JSON-RPC provider.
+    pub l1_provider: Url,
+    /// Address of LightClient contract on layer 1.
+    pub light_client_address: Address,
+    /// Transaction signing key for Ethereum
+    pub eth_signing_key: SigningKey,
+    /// Number of nodes
+    pub num_nodes: usize,
+    /// Seed to generate keys
+    pub seed: [u8; 32],
+    /// If daemon and provided, the service will run a basic HTTP server on the given port.
+    ///
+    /// The server provides healthcheck and version endpoints.
+    pub port: Option<u16>,
+}
+
+pub async fn init_stake_table(
+    config: &StateProverConfig,
+) -> StakeTable<BLSPubKey, StateVerKey, CircuitField> {
+    // We now initialize a static stake table as what hotshot orchestrator does.
```

Comment:
Why can't we reuse the code of the orchestrator?

Reply:
As the design of orchestrator seems not stable for now. And the current orchestrator cannot provide the needed information, `known_node_with_stakes` doesn't contain the Schnorr ver key.